### DISCUSSION TITLE
Return all resolved IPs for a specific domain instead of only the fastest

### DISCUSSION THREAD

**Discussion by dkadioglu** [pymumu/smartdns#863] [Q&A]
Hi,

I'm using smartdns on OpenWRT, which works great so far. However, I would like to achieve, that for specific domains all IPs are returned and not only the fastest. Is that possible?

If I understood the feature "no-speed-check" correctly, this should do it for all domains. The domain-rule option "speed-check-mode none" should do it for a specified domain. Is that right, or do I misunderstand the functionality?

What I've tried so far:
- add `speed-check-mode none ` to /etc/smartdns/custom.conf -> no change in behaviour, only one IP returned
- add `domain-rules /www.domain.de/ -speed-check-mode none` to /etc/smartdns/address.conf -> no change in behaviour, only one IP returned

I would be thankful for any help! Thanks

**Comment by PikuZheng** [comment]
I don't fully understand. Do you want to keep waiting until receive all response from upstream server and merge all results?
what if an upstream server not response (or timeout) , will it cause all DNS queries to be slow?

**Comment by dkadioglu** [comment]
Thanks for your reply.
I am not talking about asking all upstream DNS resolvers. I want to get all IPs of a resolved domain name. Maybe an example helps:

When I resolve the domain name via SmartDNS, I get
```
root@openwrt:~# drill some.domain.de
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 22989
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; some.domain.de.        IN      A

;; ANSWER SECTION:
some.domain.de.   47628   IN      A       1.2.3.4

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 2 msec
;; SERVER: ::1
;; WHEN: Sat Jan  1 13:29:48 2022
;; MSG SIZE  rcvd: 74
```

When I resolve the domain name directly (bypassing SmartDNS), I get
```
root@openwrt:~# drill some.domain.de
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 34230
;; flags: qr rd ra ; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; some.domain.de.        IN      A

;; ANSWER SECTION:
some.domain.de.   18842   IN      A       1.2.3.4
some.domain.de.   18842   IN      A       1.2.3.5

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 1 msec
;; SERVER: ::1
;; WHEN: Sat Jan  1 13:18:11 2022
;; MSG SIZE  rcvd: 70
```

As you can see, in the second run of drill bypassing SmartDNS, I get 2 IPs for some.domain.de. This is, what I want to achieve. In the best case only for specific domains, i.e. on a case by case basis, I would like to disable the mechanism of "returning only the fastest IP". As far as I understood, `speed-check-mode none` would do exactly that. Or is there a misunderstanding?

**Reply by PikuZheng** [reply]
oh I see. Clean cache file is what you need after configuring `speed-check-mode none`.

**Comment by dkadioglu** [comment]
OK, thanks. I stopped smartdns, deleted the cache file /tmp/smartdns.cache and started smartdns. However, still only a single IP. Are there specific steps necesssary to clear the cache?

**Reply by PikuZheng** [reply]
maybe the upstream of smartdns only responded to one ip. also consider the local cache, try testing with a different domain name

**Reply by dkadioglu** [reply]
OK, seems to work now. I tested both settings again and gave some hours to ensure, that all caches have been invalidated, with the following result:
- add speed-check-mode none to /etc/smartdns/custom.conf -> if applicable, multiple IPs are returned
- add domain-rules /www.domain.de/ -speed-check-mode none to /etc/smartdns/address.conf -> for the specific domain multiple IPs are returned

I'll leave this open for a couple of days to be sure. Thank you very much for leading me into the right direction regarding the existence of multiple caches.
